source/shared/core_stmt.cpp

```diff
@@ -2682,6 +2661,129 @@ bool reset_ae_stream_cursor( _Inout_ sqlsrv_stmt* stmt ) {
     return false;
 }
 
+void adjustInputPrecision( _Inout_ zval* param_z, _In_ SQLSMALLINT decimal_digits ) {
+    // 38 is the maximum precision supported for sql decimal types
+    // 6 is a composition of: 1 for '.'; 1 for sign of the number;
+    //   1 for 'e' or 'E' (scientific notation); 1 for sign of scientific exponent; 2 for length of scientific exponent
+    size_t maxDecimalLen = 38 + 6;
+    SQLSRV_ASSERT( Z_STRLEN_P( param_z ) < maxDecimalLen, "Input decimal overflow: sql decimal type only supports up to a precision of 38." );
+    std::vector<size_t> digits;
+    char* ptr = ZSTR_VAL( Z_STR_P( param_z ));
+    bool isNeg = false;
+    char scientificChar = ' ';
+    int scientificExp = 0;
+    // parse digits in param_z into the vector digits
+    if( *ptr == '+' || *ptr == '-' ){
+        if( *ptr = '-' ){
+            isNeg = true;
+        }
+        ptr++;
+    }
+    size_t numInt = 0;
+    size_t numDec = 0;
+    while( isdigit( *ptr )){
+        digits.push_back( *ptr - '0' );
+        ptr++;
+        numInt++;
+    }
+    if( *ptr == '.' ){
+        ptr++;
+        while( isdigit(*ptr) ){
+            digits.push_back( *ptr - '0' );
+            ptr++;
+            numDec++;
+        }
+    }
+    if( *ptr == 'e' || *ptr == 'E' ){
+        scientificChar = *ptr;
+    }
+    if( scientificChar != ' ' ){
+        ptr++;
+        bool isNegExp = false;
+        if( *ptr == '+' || *ptr == '-' ){
+            if( *ptr == '-' ){
+                isNegExp = true;
+            }
+            ptr++;
+        }
+        while( isdigit( *ptr )){
+            scientificExp = scientificExp * 10 + *ptr - '0';
+            ptr++;
+        }
+        SQLSRV_ASSERT( scientificExp <=38, "Input decimal overflow: sql decimal type only supports up to a precision of 38." );
+        if( isNegExp ){
+            scientificExp = scientificExp * -1;
+        }
+    }
+    SQLSRV_ASSERT( *ptr == '\0', "Invalid input decimal: Invalid character found in the decimal string." );
+    // if number of decimal is less than the exponent, that means the number is a whole number, so no need to adjust the precision
+    if(( int )numDec > scientificExp ){
+        int decToRemove = numDec - scientificExp - decimal_digits;
+        if( decToRemove > 0 ){
+            bool carryOver = false;
+            int backInd = 0;
+            // pop digits from the vector until there is only 1 more decimal place than required decimal_digits
+            while( decToRemove != 1 && !digits.empty() ){
+                digits.pop_back();
+                decToRemove--;
+            }
+            if( !digits.empty() ){
+                // check if the last digit to be popped is greater than 5, if so, the digit before is needs to round up
```

Comment:
Why don't you use some variables up there to remember the exact positions where to pop or carry over?

Reply:
can you be a bit more specified? do you mean to have booleans to keep track of which number in digits are greater than 5? I don't want to do that since we'll have to keep track of another vector of booleans, and by then you still need to ask if the boolean at that position is true or not, which isn't that much faster than after if the digit itself is greater than or equals to 5.

Reply:
Putting scientific notation aside, while parsing a number with decimals, you already know the precision you want, so you can keep track of how many decimals you have parsed so far and decided whether to continue parsing or not. If not, right then check if you need to round up the the digit that is last scanned.